bpo-10379: deprecate locale.format in lieu of locale.format_string

plusminushalf opened the pull request:
```python
>>> import locale;
>>> print(locale.format('%.0f KB', 100))
100 KB
>>> locale.format.__doc__
'Deprecated, use format_string instead.'
```

plusminushalf committed:
bpo-10379: deprecate locale.format in lue of locale.format_string

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow these steps to rectify the issue:

1. Sign the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/)
2. **Wait at least one US business day** and then check "Your Details" on [bugs.python.org](https://bugs.python.org/) to see if your account has been marked as having signed the CLA (the delay is due to a person having to manually check your signed CLA)
3. Reply here saying you have completed the above steps


Thanks again to your contribution and we look forward to looking at it!

bitdancer commented:
Thanks for working on this.  Doc changes are also needed.

bitdancer commented:
As well as tests for the deprecation message.

plusminushalf committed:
Adding tests and documentation changes

bitdancer commented:
Somehow I hadn't noticed that 'monetary' wasn't supported by format_string.  So that is actually an enhancement, in addition to the deprecation, and needs to be documented as such (versionchanged, whats new entry).  Also, format itself shouldn't be changed, *just* deprecated (that is, the deprecation message code added, but nothing else in it changed).

plusminushalf committed:
Adding version changed info for locale.format

plusminushalf commented:
@bitdancer I don't think to keep locale.format as it is and just displaying deprecated warning would be of any use. I think it is safe to call locale.format_string internally.

bitdancer commented on Doc/library/locale.rst:
```diff
@@ -374,6 +374,11 @@ The :mod:`locale` module defines the following exception and functions:
+   grouping strings.
+   Replaces :meth:`format`.
+
 
```
The description of the monetary parameter should be added to the regular docs, and the version changed phrase should just say "the monetary keyword parameter was added"

bitdancer commented:
That's the way we normally do deprecations: leave the existing code in place (so its behavior doesn't change) but point people to the preferred solution.  It would be acceptable if there were no behavior changes, but the changed tests prove that there are.

plusminushalf committed:
Deprecating but not changing the function defination

bitdancer commented on Doc/library/locale.rst:
```diff
@@ -375,8 +375,7 @@ The :mod:`locale` module defines the following exception and functions:
    .. versionchanged:: 3.7
-   Added *monetary*, if  true the conversion uses monetary thousands separator and
-   grouping strings.
+   The *monetary* keyword parameter was added.
```
We still need the description of the monetary keyword (presumably copy and pasted from the existing format entry) in the description of the format_string function.

In fact, what we really want to do is copy most of the 'format' description into format_string, since it seems wrong somehow to refer to the docs of a deprecated from from the preferred function.

bitdancer commented on Lib/test/test_locale.py:
```diff
@@ -197,6 +198,13 @@ def test_padding(self):
         self._test_format("%-10.f", 4200, grouping=0, out='4200'.ljust(10))
 
+    def test_format_deprecation(self):
+        with warnings.catch_warnings(record=True) as w:
```
You can use assertWarns here instead.

bitdancer commented on Doc/library/locale.rst:
```diff
@@ -366,11 +369,15 @@ The :mod:`locale` module defines the following exception and functions:
 
+   .. versionchanged:: 3.7
+   The *monetary* keyword parameter was added.
+   Replaces :meth:`format`.
```
I don't think the 'replaces' sentence is needed.  The deprecated on format covers that.

bitdancer reviewed (changes requested):
Thanks.  Almost there :)

It would also be good to have an entry for the What's New document for 3.7 for the addition of the monetary parameter to format_string, as well as the deprecation of format.  I haven't looked; if the 3.7 what's new hasn't been organized yet, just stick in a placeholder sentence with a reference to the bpo issue number.

plusminushalf committed:
Adding documentation inside what's new in 3.7

Fixed recommended CR changes

bitdancer commented on Doc/library/locale.rst:
```diff
@@ -366,11 +369,22 @@ The :mod:`locale` module defines the following exception and functions:
    locale settings into account.
 
+   .. versionchanged:: 3.7
+   The *monetary* keyword parameter was added.
```
Missing indentation.

bitdancer commented on Doc/whatsnew/3.7.rst:
```diff
@@ -158,6 +158,11 @@ API and Feature Removals
   function instead.
 
+* Deprecated :meth:`format` from :mod:`locale`, use the :meth:`format_string`
+  instead. Added another argument *monetary* in :meth:`format_string` of
```
format hasn't been removed, so the deprecation should go in the deprecation section.  The feature addition of the monetary keyword should go in the 'improved modules' section.

bitdancer reviewed (changes requested):
Thanks for sticking with this :)

plusminushalf committed:
Adding whatsnew documentation in proper segments

plusminushalf committed:
Merge branch 'master' of github.com:python/cpython into fix-bpo-10379

plusminushalf committed:
fixing documentation bug

bitdancer commented on Doc/whatsnew/3.7.rst:
```diff
@@ -93,6 +93,10 @@ New Modules
 Improved Modules
 ================
 
+* Added another argument *monetary* in :meth:`format_string` of :mod:`locale`.
```
There should be a subsection header before this line for the locale module (the modules with enhancements are listed in this section in alphabetical order).

plusminushalf committed:
Adding locale subsection

plusminushalf commented:
@bitdancer is it good to go now?

bitdancer commented:
Yes, I think so.  I'd like to do a final overall review before I approve it, though.  With any luck I'll do that tonight.

plusminushalf commented:
Thanks! :)

plusminushalf committed:
Merge branch 'master' of github.com:python/cpython into fix-bpo-10379

plusminushalf commented:
Hey, @bitdancer have you reviewed this?

bitdancer commented:
I did...and I feel bad that I have some more changes to suggest, so I was going to make them myself and submit it to the PR, but I haven't actually figured out how to do that yet.

plusminushalf commented:
@bitdancer well I think you can suggest me the changes I'll do my best to figure them out.

bitdancer commented on Doc/library/locale.rst:
```diff
@@ -365,12 +365,26 @@ The :mod:`locale` module defines the following exception and functions:
    For whole format strings, use :func:`format_string`.
 
+   .. deprecated:: 3.7
+      Use :meth:`format_string` instead
```
So looking at this again, I think what we should do is move 'format' under 'format_string', and change *its* body to say "works like format_string but only accepts a single format specification.  That followed by the deprecation notice will make it a lot less likely anyone will use it :)

bitdancer commented on Doc/whatsnew/3.7.rst:
```diff
@@ -97,6 +97,14 @@ New Modules
+If *monetary* is true, the conversion uses monetary thousands separator and
+grouping strings.
+(Contributed by Garvit in :issue:`10379`.)
+
```
Small nit: our convention appears to be to put the Contributed by as part of the paragraph when there's only one paragraph.  Ends up formatted the same in html, but might as well make the ReST consistent as well.

bitdancer commented on Lib/locale.py:
```diff
@@ -181,12 +182,13 @@ def _strip_padding(s, amount):
+    """Deprecated, use format_string instead."""
+    warnings.warn(
+        "This method will be removed in future versions.  "
+        "Use 'locale.format_string()' instead.",
```
I think this should read "This method will be removed in a future version of Python."

bitdancer commented on Lib/locale.py:
```diff
@@ -217,10 +219,12 @@ def _format(percent, value, grouping=False, monetary=False, *additional):
     """Formats a string in the same way that the % formatting would use,
     but takes the current locale into account.
-    Grouping is applied if the third parameter is true."""
+    Grouping is applied if the third parameter is true.
```
As long as we are editing this, I think there should be a blank line after the first sentence, before the 'Grouping' sentence.

bitdancer commented on Lib/locale.py:
```diff
@@ -262,7 +266,7 @@ def currency(val, symbol=True, grouping=False, international=False):
                          "the 'C' locale.")
 
-    s = format('%%.%if' % digits, abs(val), grouping, monetary=True)
+    s = format_string('%%.%if' % digits, abs(val), grouping, monetary=True)
```
Since we *know* there's only one format string here, I think this should be a call to _format.

bitdancer commented on Lib/locale.py:
```diff
@@ -298,7 +302,7 @@ def currency(val, symbol=True, grouping=False, international=False):
 def str(val):
     """Convert float to string, taking the locale into account."""
-    return format("%.12g", val)
+    return format_string("%.12g", val)
```
As above, I think this should be _format.

plusminushalf committed:
Refactoring Doc, using _format inisde locale module

Refactoring Doc, moving format doc below format_string,
this is to make sure people have lesser probability of using
format in future.
Also using _format method inside locale module, since we know
there is only one format function that is _format

bitdancer commented:
There was no need to move format in the source file, but it doesn't hurt, either.

plusminushalf commented:
Hey @bitdancer is there anything left to get this merged?

bitdancer commented:
Just adding the Misc/NEWS entry.  I'm not sure what status is on how to do that currently, but I think it is still manual edit of that file.  I haven't had time yet to learn how to update a PR before merge, so if you add the entry, I'll hit merge :)

plusminushalf committed:
Adding entry to Misc/NEWS about changes

bitdancer commented on Misc/NEWS:
```diff
@@ -364,6 +364,8 @@ Library
 - bpo-29534: Fixed different behaviour of Decimal.from_float()
   for _decimal and _pydecimal. Thanks Andrew Nester.
 
+- bpo-10379: Deprecate locale.format in lue of locale.format_string
```
Google says it's spelled 'lieu'.  Also the entry should note that the monetary argument was added to format_string.  So, how about "locale.format_string now supports the 'monetary' keyword argument, and locale.format is deprecated."

plusminushalf committed:
CR fixes
